execute/tokendata/cctp/v2/observer.go

```diff
@@ -0,0 +1,522 @@
+
+// getMessageTokenDataForSourceChain takes a collection of CCIP Messages for a source chain, collects the messages'
+// TokenAmounts that are CCTP v2 transfers, fetches the attestations for these CCTP v2 transfers, and converts them to
+// MessageTokenData.
+//
+// Transaction hashes are collected and used to batch-fetch CCTP messages from the attestation service, as multiple
+// tokens within the same transaction will share the same attestation data. The function returns a map with the same
+// keys as the input ccipMessages map, ensuring each input message gets corresponding token data.
+func getMessageTokenDataForSourceChain(
+	ctx context.Context,
+	lggr logger.Logger,
+	sourceChain cciptypes.ChainSelector,
+	ccipMessages map[cciptypes.SeqNum]cciptypes.Message,
+	supportedPoolsBySelector map[cciptypes.ChainSelector]string,
+	attestationEncoder AttestationEncoder,
+	attestationClient CCTPv2AttestationClient,
+	metricsReporter MetricsReporter,
+) map[cciptypes.SeqNum]exectypes.MessageTokenData {
+	// Add sourceChain to logger context for all logs in this function
+	lggr = logger.With(lggr, "sourceChain", sourceChain)
+
+	// Step 1: Check if source chain has CCTP v2 support configured
+	// If not, all tokens are marked as not supported
+	cctpV2EnabledTokenPoolAddress, ok := supportedPoolsBySelector[sourceChain]
+	if !ok {
+		return notSupportedMessageTokenData(ccipMessages)
+	}
+
+	// Step 2: Extract and decode CCTP v2 token data payloads from CCIP messages
+	sourceTokenDataPayloads := make(map[cciptypes.SeqNum]map[int]SourceTokenDataPayload)
+	for seqNum, ccipMessage := range ccipMessages {
+		sourceTokenDataPayloads[seqNum] = getSourceTokenDataPayloads(ccipMessage, cctpV2EnabledTokenPoolAddress)
+	}
+
+	// Step 3: Validate that all CCTP v2 tokens have the same source domain ID
+	// This is required because we need to query Circle's API with a single domain ID
+	sourceDomainID, err := getSourceDomainID(lggr, sourceChain, sourceTokenDataPayloads)
+	if err != nil {
+		// Return error tokens for all CCTP v2 tokens, others remain not supported
+		return errorMessageTokenData(err, ccipMessages, sourceTokenDataPayloads)
+	}
+
+	// Step 4: Extract unique transaction hashes from messages with CCTP v2 tokens
+	txHashes := getTxHashes(lggr, sourceTokenDataPayloads, ccipMessages)
+
+	// Step 5: Fetch CCTP v2 messages and attestations from Circle's API
+	// Query the attestation service for all collected transaction hashes
+	// Each response contains the message data and attestation needed for on-chain minting
+	cctpV2Messages, err := getCCTPv2Messages(ctx, lggr, attestationClient, sourceChain, sourceDomainID, txHashes)
+	if err != nil {
+		// Return error tokens for all CCTP v2 tokens, others remain not supported
+		return errorMessageTokenData(err, ccipMessages, sourceTokenDataPayloads)
+	}
+
+	// Step 6: Match each SourceTokenDataPayload to its corresponding CCTP v2 Message
+	tokenIndexToCCTPv2Message := matchCCTPv2MessagesToSourceTokenDataPayloads(
```

Comment:
Is it possible to have this information on the cctp message details be retrieved from an event log rather than from the source token data payload? This would save us a LOT of space on the amount of bytes being sent between chains since most of the fields in the new struct are not used on destination chain?